### PR TITLE
Added support for discovering and reading Virgo FFL files

### LINKED ISSUE #748: Add support for Virgo FFL files
At Virgo, the way for the end user to find off-line data is by pointing your Matlab/Python/dataDisplay to a "FFL file" (raw.ffl, trend.ffl, ...). This is a pretty basic text format, where every line lists a file name, gps_start and duration. Reading data via an FFL files is built natively into the FrameL library. As far as I know this doesn't work with gwpy yet. It would be good to implement this, to make gwpy work at the Virgo site.

I will later attach some sample ffl file and possibly some small gwf files.

### PR DESCRIPTION
**Pull request by duncanmmacleod**
This PR adds support for auto-discovery of data on the Virgo file system (Cascina) by supporting reading FFL files, and auto-discovery of those files under the `/virgoData` partition.

Fixes #748.

cc @basswinkels

### CODE REVIEW

**Comment by basswinkels** [comment]
Thanks for working on this! I am trying to test and understand all the new functionality, I will get back to you in the next days.

**Comment by duncanmmacleod** [comment]
@basswinkels, there is a catch to using this on the Virgo filesystem: matching channels to a frametype requires the `frameCPP`, which isn't installed by default, and isn't (yet) available from Conda. However, discovering and reading files when the frametype is *known* should work now, e.g.:

```python
>>> TimeSeries.get('V1:Hrec_hoft_16384Hz', <start>, <end>, frametype='V1Online')
```

**Comment by basswinkels** [comment]
@duncanmmacleod 
Things I did roughly:
- cloned your repository and applied the PR (still new to git, so this took a while)
- created a new env with `conda create --name gwpy_ffl python=2.7 pip ipython` and activated it.
- installed the patched gwpy inside the env with `pip install -e ~/git/gwpy/`
- enabled lalsuite (for frame access) with `source  /virgoApp/lalsuite/v6r48/$UNAME/etc/lalsuite-user-env.sh`

With this, the following seems to work fine:
`d = TimeSeries.read('/virgoData/ffl/raw.ffl', 'V1:INJ_IMC_TRA_DC', start=gs, end=ge)`
apart from the scaling issue of #749. This is anyhow very cool, and would allow further testing of gwpy in some of my noise tools.

The following things don't seem to work yet:
```
In [10]: d = TimeSeries.get('V1:INJ_IMC_TRA_DC', gs, ge, frametype='raw')
[...]
ImportError: No module named nds2

In [11]: d = TimeSeries.find('V1:INJ_IMC_TRA_DC', gs, ge, frametype='raw')
[...]
ImportError: No module named gwdatafind
```
Is that expected? I guess I could just install those, but especially nds2 seems LIGO only.

After `pip install gwdatafind` , both commands give a `IOError: [Errno 22] Invalid argument`, I am probably doing something wrong.

**Comment by duncanmmacleod** [comment]
@basswinkels, thanks for the notices. can you post the full traceback you get from your `d = TimeSeries.find('V1:INJ_IMC_TRA_DC', gs, ge, frametype='raw')` call _after_ you have installed `gwdatafind`? You shouldn't get any errors in that case.

Can you also post the `gs` and `ge` you are using, so that I can reproduce your issues?

**Comment by basswinkels** [comment]
@duncanmmacleod see attached:
[ffl_traceback.txt](https://github.com/gwpy/gwpy/files/2416846/ffl_traceback.txt)

File: gwpy/io/datafind.py [hunk 1]
```diff
@@ -40,32 +48,299 @@
...
+        self.ffldir = ffldir or self._get_ffl_dir()
+        self.paths = {}
+        self.cache = {}
+        self._find_paths()
+
+    # -- utilities ------------------------------
+
+    @staticmethod
+    def _get_ffl_dir():
+        return os.path.join(os.environ['VIRGODATA'], 'ffl')
```
[the remark this answers is not in the document]
Comment (duncanmmacleod) [reply gwpy/io/datafind.py]: thanks for this, I will implement `FFLPATH` (following convention of `MATLABPATH` and `PYTHONPATH`) as the default, falling back to `${VIRGODATA}/ffl`.

File: gwpy/io/cache.py [hunk 2]
```diff
@@ -64,42 +68,74 @@
 
 # -- cache I/O ----------------------------------------------------------------
 
-def read_cache(lcf, coltype=LIGOTimeGPS):
-    """Read a LAL-format cache file
+def read_cache(cachefile, coltype=LIGOTimeGPS):
+    """Read a LAL- for FFL-format cache file as a list of file paths
```
[the remark this answers is not in the document]
Comment (duncanmmacleod) [reply gwpy/io/cache.py]: thanks

**Comment by duncanmmacleod** [comment]
> @duncanmmacleod see attached:
> [ffl_traceback.txt](https://github.com/gwpy/gwpy/files/2416846/ffl_traceback.txt)

@basswinkels, as you discovered, the `frametype` is the extensionless name of the FFL file. I think I fixed the `IOError` you were seeing (50c9d62a41421cb62051f278ead9415448aa6ae9), can you try your tests again?

**Comment by basswinkels** [comment]
Hi @duncanmmacleod 
with the latest fix (and gwdatafind installed), the following now all give the same result: 
```
TimeSeries.read('/virgoData/ffl/raw.ffl', 'V1:INJ_IMC_TRA_DC', start=gs, end=ge)
TimeSeries.find('V1:INJ_IMC_TRA_DC', gs, ge, frametype='raw')
TimeSeries.get('V1:INJ_IMC_TRA_DC', gs, ge, frametype='raw')
```
 :+1:

While playing a bit with the FflConnection object, I found a likely bug: [find_latest_bug.txt](https://github.com/gwpy/gwpy/files/2420201/find_latest_bug.txt)
I think the problem is in [this line](https://github.com/gwpy/gwpy/blob/af086a7782c7ef3d3519f1b8fa0a5ce55fab7457/gwpy/io/datafind.py#L243). It seems that `self.cache = mtime, <list of cache entries>`, so you likely need to do `cache = self.cache[(site, frametype)][1][-1:]`. You might want to do a `_read_ffl_cache()` to make sure that the cache is up to date, otherwise you cannot be sure that it is really the latest file (e.g. raw_full.ffl is modified every 10 seconds).

**Comment by duncanmmacleod** [comment]
@basswinkels, 3091bbf fixes that bug in `find_latest`.

**Comment by duncanmmacleod** [comment]
@basswinkels, are you happy enough with this implementation? I would be happy to receive subsequent issues reporting any bugs we find later down the line.

**Comment by basswinkels** [comment]
@duncanmmacleod sorry, I was busy with working in Cascina and travel. I think it is fine as is, but would like to do play a bit more with it myself. I will get back to you before monday.

**Comment by basswinkels** [comment]
Hi @duncanmmacleod 

I found one more issue: trying to read data from trend.ffl fails:
```
In [40]: g = 1222593918
In [41]: d = TimeSeries.read('/virgoData/ffl/trend.ffl', 'V1:DAQ_FFL_TREND_LATENCY', start=g, end=g+600)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-41-03e6b8ef73b1> in <module>()
----> 1 d = TimeSeries.read('/virgoData/ffl/trend.ffl', 'V1:DAQ_FFL_TREND_LATENCY', start=g, end=g+600)

/users/swinkels/git/gwpy/gwpy/timeseries/core.py in read(cls, source, *args, **kwargs)
    270         -----"""
    271         from .io.core import read as timeseries_reader
--> 272         return timeseries_reader(cls, source, *args, **kwargs)
    273 
    274     def write(self, target, *args, **kwargs):

/users/swinkels/git/gwpy/gwpy/timeseries/io/core.py in read(cls, source, *args, **kwargs)
     35         source = preformat_cache(source, *args[1:],
     36                                  start=kwargs.get('start'),
---> 37                                  end=kwargs.get('end'))
     38 
     39     # get join arguments

/users/swinkels/git/gwpy/gwpy/timeseries/io/cache.py in preformat_cache(cache, start, end)
     56         cache = read_cache(cache)
     57     cache = type(cache)(cache)  # copy cache
---> 58     cache.sort(key=file_segment)  # sort
     59 
     60     # get timing

/users/swinkels/git/gwpy/gwpy/io/cache.py in file_segment(filename)
    309         base = os.path.basename(filename)
    310         try:
--> 311             _, _, start, end = base.split('-')
    312         except ValueError as exc:
    313             exc.args = ('Failed to parse {0!r} as LIGO-T050017-compatible '

ValueError: Failed to parse u'V-914803200-01-Jan-2009-01h00-48F.T' as LIGO-T050017-compatible filename
```

The problem seems that some files in trend.ffl older than early 2015 don't have a name that follows the conventions:
```
/data/archive_trenddata//2010/V-958780800-25-May-2010-02h00-48F.T   958780800.000000   63000.000000   0.000000  0.000000
/data/archive_trenddata//2010/T-958843800-48F.gwf   958843800.000000   23400.000000   0.000000  0.000000
...
/data/archive_trenddata//2015/T-1108857600-48F.gwf   1108857600.000000   86400.000000   0.000000  0.000000
/data/archive_trenddata//2015/T-1108944000-48F.gwf   1108944000.000000   45000.000000   0.000000  0.000000
/data/archive_trenddata//2015/V-trend-1109116800-86400.gwf   1109116800.000000   86400.000000   0.000000  0.000000
/data/archive_trenddata//2015/V-trend-1109203200-86400.gwf   1109203200.000000   86400.000000   0.000000  0.000000
```

It seems that your code is just looking at the file name to determine the segment, and ignores the times written in the 2nd and 3rd column of the ffl file. So this is not a real bug, but more a case of us not following the standard in the past. I see 3 options:

1.  rewrite the code so that it uses the numbers written in the file instead of parsing the filename
2.  keep getting the segment from the filename, but silently (or with a single warning) ignore all lines for which the file does not follow the standard. This means we cannot read very old data, which is not the biggest issue.
3. ask our computing people to rename all the old files.

What do you suggest? I think 1) would be the 'correct ' solution, but 2) and/or 3) might be more practical.

**Comment by basswinkels** [comment]
And two smaller questions, which are hard to figure out without a good understanding of FrameCPP or lalframe:
- is the reading of 'procdata'  channels supported? (channels are stored as either ADCdata, Procdata or Simdata structures). A quick test suggests it does.

- how are vectors with missing samples handled? When some samples are missing in the middle of a vector at Virgo, the `next` field in the adc structure points to a separate 8-bit vector that acts like a mask. Does LIGO use the same convention, or do  you just use NaNs (which would be my preferred option)?

**Comment by duncanmmacleod** [comment]
> It seems that your code is just looking at the file name to determine the segment, and ignores the times written in the 2nd and 3rd column of the ffl file. So this is not a real bug, but more a case of us not following the standard in the past. I see 3 options:
> 
> 1. rewrite the code so that it uses the numbers written in the file instead of parsing the filename
> 2. keep getting the segment from the filename, but silently (or with a single warning) ignore all lines for which the file does not follow the standard. This means we cannot read very old data, which is not the biggest issue.
> 3. ask our computing people to rename all the old files.
> 
> What do you suggest? I think 1) would be the 'correct ' solution, but 2) and/or 3) might be more practical.

I will attempt to implement (1), which I agree is the 'correct' solution.

> * is the reading of 'procdata' channels supported? (channels are stored as either ADCdata, Procdata or Simdata structures). A quick test suggests it does.

Yes, all three types are handled.

> * how are vectors with missing samples handled? When some samples are missing in the middle of a vector at Virgo, the next field in the adc structure points to a separate 8-bit vector that acts like a mask. Does LIGO use the same convention, or do you just use NaNs (which would be my preferred option)?

I have never encountered missing samples in an `FrVect`. If you can point me at a frame reading option that breaks, I will attempt to work out what to do. If the Virgo solution is compliant with the GWF spec, then we can probably just use that as implemented in frameCPP. If it is compliant, but not implemented in frameCPP, I can work with Ed Maros to get it implemented - he has been very responsive to these sorts of requests in the past.

**Comment by duncanmmacleod** [comment]
ping @basswinkels, can you comment on my query about missing samples above?

**Comment by basswinkels** [comment]
Hi @duncanmmacleod, sorry for the delay, I was travelling for a week and had to catch with some urgent work after that, and the missing sample is unfortunately not something straightforward. I finally have some time now.

Regarding indicating missing samples with a separate mask vector: I cannot find anything about this in the official frame specification, I only found it mentioned casually in the FrameL manual, but that was regarding vectors returned by the library, not about how this is actually stored on disk. I am sure that this is done in a similar way (i.e. for frames with missing samples, the adcdata->data->next contains a separate FrVect that acts as a mask), but this might be Virgo only. Do you have an idea how missing samples are handled at LIGO?

I played a bit with reading data around a known gap in a channel in trend.ffl, see this plot:
missing_data

Note that although the gwf files are 86400 seconds long, the frames inside this gwf have a length of 1800 seconds.

Using virgotools.getChannel (which is based directly on the function FrFileIGetVectDN from FrameL, which handles merging data than spans multiple gwf files), I read the missing samples as zero. Reading across the gap, the gap is filled with zeros without giving an error. It should be possible (but this is not implement yet) to substitute the zeros by e.g. NaNs based on the separate mask vector. It only gives me an error for a missing channel in case the start time falls within the 1800 seconds for which the channel is completely missing.

Using TimeSeries.read (on a custom ffl file which excludes the illegal file names), I also read zeros for missing samples (which makes sense, since this is how they are stored on disk). I can however not read across the gap, this raises an error "RuntimeError: Internal function call failed: Wrong name"
See attached file for the complete traceback:
[data_with_gaps.txt](https://github.com/gwpy/gwpy/files/2511554/data_with_gaps.txt)

My suggestion would be to leave the handling of missing samples out of gwpy for now, since I don't even know what would be the best way to implement it. I will try to implement that first for my own code in virgotools, before I would ask to implement something similar in gwpy.

My only suggestion would be to make the error message a bit less cryptic. I guess changing the class of the exception is difficult for backwards compatibility, but it would be helpful for debugging if the error message contains something like "could not find channel XX in frame YY.gwf at gps = ZZ"

For me the only blocking issue is the one mentioned above about about getting the gwf duration from the ffl file instead of from the file name, but other than that I would suggest to merge this and let some more users test it.

**Comment by basswinkels** [comment]
@duncanmmacleod I can confirm that I can now read channels from all of the important ffl files. Even reading a historic channel from 2008, which is stored in a gwf file with a wrong name, seems to work fine. Thanks for fixing this!

**Pull request merged**